File: 2025_2026_Predictive Modelling_Forecasting/Python_Code_en_Data/Py/wine_script.py
#%%
import matplotlib.pyplot as plt
import pandas as pd
from statsmodels.tsa.statespace.sarimax import SARIMAX

wine_sales = pd.read_csv('../data/wine_sales.csv', parse_dates=['date'])
plt.figure(figsize=(12,8))
plt.title('Historical wine sales in Australia since 1980')
plt.plot( wine_sales['date'], wine_sales['sales'])
plt.show()

###--- Training the model ---###

n_obs = len(wine_sales['sales'])
n_holdout = 12

# Convert sales into a time series with monthly frequency
wine_sales.set_index('date', inplace=True)
wine_ts = wine_sales['sales'].asfreq('MS')  # Monthly start frequency

# Train-test split
train = wine_ts.iloc[:-n_holdout]
test = wine_ts.iloc[-n_holdout:]

# Fit SARIMA model manually
# Using SARIMA(1,1,1)(1,1,1,12) as a reasonable starting point for seasonal monthly data

full_arima = SARIMAX(wine_ts, order=(1,1,1), seasonal_order=(1,1,1,12)).fit()

# Extract fitted values for train and test
train_res = pd.DataFrame({
    'date': train.index,
    'sales': train.values,
    'arima': full_arima.fittedvalues.iloc[:len(train)]
})

test_res = pd.DataFrame({
    'date': test.index,
    'sales': test.values,
    'arima': full_arima.fittedvalues.iloc[len(train):]
})


###--- Evaluate model performance ---###

# Plot in-sample fit
plt.figure(figsize=(12, 6))
plt.plot(train_res['date'], train_res['sales'], color='red', marker='o', label='Real Sales')
plt.plot(train_res['date'], train_res['arima'], color='blue', marker='o', label='Fitted Sales')
plt.title('Real sales (red) vs fitted sales (blue)')
plt.legend()
plt.show()

# Plot out-of-sample predictions
plt.figure(figsize=(12, 6))
plt.plot(test_res['date'], test_res['sales'], color='red', marker='o', label='Real Sales')
plt.plot(test_res['date'], test_res['arima'], color='blue', marker='o', label='Predicted Sales')
plt.title('Real sales (red) vs predicted sales (blue)')
plt.legend()
plt.show()

wine_ts = pd.Series(
    data=wine_sales["sales"].values,
    index=pd.date_range(start="1980-01-01", periods=len(wine_sales), freq="M")
)



# %%

File: 2025_2026_Predictive Modelling_Forecasting/R_Code/wine_script.R
###--- Setup ---###

library(forecast)
library(ggplot2)
library(lubridate)

load('wine_sales.RData')


###--- Preliminary analysis ---###

# plot of wine data
ggplot(wine_sales, aes(x = date, y = sales)) + 
  geom_line(col = 'blue', size = 1) + 
  ggtitle('Historical wine sales in Australia since 1980' )


###--- Training the model ---###

n_obs <- length(wine_sales$sales)
n_holdout <- 12

# transform original sales into time series (ts) object
# since we are interested in modeling seasonality, it is crucial
# to include the frequency. In our case, we assume that a pattern 
# repeats itself every year -> frequency = 12 months. If a pattern 
# occurs every week, and we have daily data, then the frequency is 
# 1 week -> 7 days. If a patern repeats itself yearly, then the 
# frequency is 365 days if your observations are measured daily. There 
# is a problem when you have a leap year (schrikkeljaar) or if some days
# are not recorded, for example due to holidays. For this reason, it 
# is NOT recommended to use frequency = 365
wine_ts <- ts(wine_sales$sales, frequency = 12, start = c(1980,1))

# define the indices of our train sample
train_index <- 1:(n_obs - n_holdout)

# create train and test sample
train <- ts(wine_sales$sales[train_index], frequency = 12, start = c(1980, 1))

# extract last date
end_date <- wine_sales$date[n_obs]
# transform date into vector containing year and month
end <- c(year(end_date), month(end_date))
test <- ts(wine_sales$sales[-train_index], frequency = 12, end = end)

## The code below does not work as you would expect; the result is a vector
## and not a ts_object. This is problematic when you want to model seasonality
## since the frequency is necessary. Use the approach shown above or the 
## windows() function to subset time series.
# train <- wine_ts[train_index] # <- wrong
# test <- wine_ts[-train_index] # <- wrong


# find arima model that leads to lowest AIC value
fit_arima <- auto.arima(train, ic = 'aic')

# now we use the Arima() function to obtain the predictions of the
# out of sample observations. Note that since we pass the model as
# an argument, we do NOT train on the new dataset.
full_arima <- Arima(wine_ts, model = fit_arima)

# This does not work since you need at least 12 months of data due
# to the seasonality.
#pred_arima <- forecast(test, h = n_holdout, model = fit_arima)
#pred_ar <- forecast(test, h = n_holdout, model = fit_ar)


###--- Evaluate model performance ---###

train_res <- data.frame(date = wine_sales$date[train_index],
                        sales = wine_sales$sales[train_index],
                        arima = full_arima$fitted[train_index])

test_res <- data.frame(date = wine_sales$date[-train_index],
                       sales = wine_sales$sales[-train_index],
                       arima = full_arima$fitted[-train_index])


# plot of the in sample model fit (train set)
ggplot(train_res, aes(x = date)) + 
  geom_line(aes(y = sales), col = 'red', size = 0.65) + 
  geom_line(aes(y = arima), col = 'blue', size = 0.65) +
  geom_point(aes(y = sales), col = 'red', size  = 1) +
  geom_point(aes(y = arima), col = 'blue', size = 1) + 
  ggtitle('Real sales (red) vs fitted sales (blue).')

# plot of the out-of-sample model predictions (test set)
ggplot(test_res, aes(x = date)) + 
  geom_line(aes(y = sales), col = 'red', size = 0.65) + 
  geom_line(aes(y = arima), col = 'blue', size = 0.65) +
  geom_point(aes(y = sales), col = 'red', size  = 1) +
  geom_point(aes(y = arima), col = 'blue', size = 1) + 
  ggtitle('Real sales (red) vs predicted sales (blue).')
  


